pkg/apiserver/rest/dualwriter_mode3_test.go

```diff
@@ -1,72 +1,348 @@
+	}
+	tests :=
+		[]testCase{
+			{
+				name:  "creating an object only in the unified store",
+				input: exampleObj,
+				setupLegacyFn: func(m *mock.Mock, input runtime.Object) {
+					m.On("Create", mock.Anything, input, mock.Anything, mock.Anything).Return(exampleObj, nil)
+				},
+				setupStorageFn: func(m *mock.Mock) {
+					m.On("Create", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(exampleObj, nil)
+				},
+			},
+			{
+				name:  "error when creating object in the unified store fails",
+				input: failingObj,
+				setupLegacyFn: func(m *mock.Mock, input runtime.Object) {
+					m.On("Create", mock.Anything, failingObj, mock.Anything, mock.Anything).Return(nil, errors.New("error"))
+				},
+				wantErr: true,
+			},
+		}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			l := (LegacyStorage)(nil)
+			s := (Storage)(nil)
+			m := &mock.Mock{}
+
+			ls := legacyStoreMock{m, l}
+			us := storageMock{m, s}
+
+			if tt.setupLegacyFn != nil {
+				tt.setupLegacyFn(m, tt.input)
+			}
+			if tt.setupStorageFn != nil {
+				tt.setupStorageFn(m)
+			}
+
+			dw := NewDualWriter(Mode3, ls, us, p)
+
+			obj, err := dw.Create(context.Background(), tt.input, func(context.Context, runtime.Object) error { return nil }, &metav1.CreateOptions{})
+
+			if tt.wantErr {
+				assert.Error(t, err)
+				return
+			}
+
+			acc, err := meta.Accessor(obj)
+			assert.NoError(t, err)
+			assert.Equal(t, acc.GetResourceVersion(), "1")
+			assert.NotEqual(t, obj, anotherObj)
+		})
+	}
+}
+
+func TestMode3_Get(t *testing.T) {
+	type testCase struct {
+		setupStorageFn func(m *mock.Mock, name string)
+		name           string
+		input          string
+		wantErr        bool
+	}
+	tests :=
+		[]testCase{
+			{
+				name:  "get an object only in unified store",
+				input: "foo",
+				setupStorageFn: func(m *mock.Mock, name string) {
+					m.On("Get", mock.Anything, name, mock.Anything).Return(exampleObj, nil)
+				},
+			},
+			{
+				name:  "error when getting an object in the inified store fails",
+				input: "object-fail",
+				setupStorageFn: func(m *mock.Mock, name string) {
+					m.On("Get", mock.Anything, name, mock.Anything).Return(nil, errors.New("error"))
+				},
+				wantErr: true,
+			},
+		}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			l := (LegacyStorage)(nil)
+			s := (Storage)(nil)
+			m := &mock.Mock{}
+
+			ls := legacyStoreMock{m, l}
+			us := storageMock{m, s}
+
+			if tt.setupStorageFn != nil {
+				tt.setupStorageFn(m, tt.input)
+			}
+
+			p := prometheus.NewRegistry()
+			dw := NewDualWriter(Mode3, ls, us, p)
+
+			obj, err := dw.Get(context.Background(), tt.input, &metav1.GetOptions{})
+
+			if tt.wantErr {
+				assert.Error(t, err)
+				return
+			}
+
+			us.AssertNotCalled(t, "Get", context.Background(), tt.name, &metav1.GetOptions{})
+
+			assert.Equal(t, obj, exampleObj)
+			assert.NotEqual(t, obj, anotherObj)
+		})
+	}
+}
+
+func TestMode3_List(t *testing.T) {
```

Comment:
Looks like we need to improve testing here because this test wouldn't catch that we were listing from Legacy instead of Storage.